### PR TITLE
Issue #376 fix

### PR DESCRIPTION
**Pull request by akotlar**
### MDL version: 
1.1 – 1.3
### Browser: 
Chrome, Firefox
### Browser version: 
Chrome 51.0.2704.103 , Firefox 47.0 
### Operating system: 
OSX Yosemite
### Operating system version: 
10.10.5
### URL, if applicable 
[Plnk](http://plnkr.co/edit/p1T5fRDDPbqnhKfyk0Ky?p=info)
[Youtube Demo]( https://youtu.be/kyUK9N-1wA8)

#### What steps will reproduce the problem:
1. Tap on any mdl button, using a Macbook Pro touchpad

#### What is the expected result?
Ripple Animation

#### What happens instead of that? 
Nothing, and on occasion, short ripple

#### Please provide any other information below, and attach a screenshot if possible.
Demo of this is provided

### CODE REVIEW

**Comment by akotlar** [comment]
These commits were authored by me. Not sure what's happened with this.

**Comment by akotlar** [comment]
Jeez, I see what happened. For some reason I merged a Readme commit. Ok, I will re-do this.

File: src/ripple/ripple.js [hunk 1]
```diff
@@ -109,25 +110,52 @@
...
+      if (this.timeout) {
+        window.clearTimeout(this.timeout);
+      }
+
+      // We've just clicked, clear past upHandler_ call record
+      this.rippleElement_.mouseUp = false;
+      // Indicate to upHandler_ that ripple hasn't run for this.Constant_.DURATION
+      this.rippleElement_.animationCompleted = false;
+
+      window.requestAnimationFrame(function() {
```
Comment (sompylasar) [inline src/ripple/ripple.js]: `requestAnimationFrame` changed to `window.requestAnimationFrame` for no reason.

File: src/ripple/ripple.js [hunk 2]
```diff
@@ -109,25 +110,52 @@
...
+        this.animFrameHandler();
+
+        // Allow the animation to complete, even for tap events, which may
+        // trigger multiple mouseup with final event.detail == 1
+        // Ex: mac trackpad tap triggers 2 consecutive event.detail == 0 events
+        // followed by an event.detail=1 mouseup event
+        // we set the animation timeout inside the frame request to
+        // prevent animation flashing during long ui-blocking operations
+        this.timeout = window.setTimeout(function() {
```
Comment (sompylasar) [inline src/ripple/ripple.js]: Probably `setTimeout` instead of `window.setTimeout`, similarly to `requestAnimationFrame`.

File: src/ripple/ripple.js [hunk 3]
```diff
@@ -109,25 +110,52 @@
...
+          // Check if use holding click event (or touchend not yet called)
+          if (this.rippleElement_.mouseUp) {
+            window.requestAnimationFrame(function() {
+              this.rippleElement_.classList.remove(this.CssClasses_.IS_VISIBLE);
+            }.bind(this));
```
Comment (sompylasar) [inline src/ripple/ripple.js]: Better to create the bound functions once, not each time (here and below).
And functions created by `.bind()` were noted to be less performant to call than the ones that capture `this` via closure.

**Comment by akotlar** [comment]
Guys, my apologies, I will submit an update in just a minute. This was my first pull request into your codebase, and I didn't get it right. The setTimeout and requestAnimation frame were not my changes I believe, just merging different versions issue

File: src/ripple/ripple.js [hunk 4]
```diff
@@ -109,25 +110,52 @@
       }
       this.setRippleXY(x, y);
       this.setRippleStyles(true);
-      requestAnimationFrame(this.animFrameHandler.bind(this));
+
+      if (this.timeout) {
+        window.clearTimeout(this.timeout);
```
Comment (sompylasar) [inline src/ripple/ripple.js]: Probably `clearTimeout` instead of `window.clearTimeout`, similarly to `requestAnimationFrame`.

File: src/ripple/ripple.js [hunk 5]
```diff
@@ -109,25 +110,52 @@
...
-    // Don't fire for the artificial "mouseup" generated by a double-click.
-    if (event && event.detail !== 2) {
-      // Allow a repaint to occur before removing this class, so the animation
-      // shows for tap events, which seem to trigger a mouseup too soon after
-      // mousedown.
-      window.setTimeout(function() {
-        this.rippleElement_.classList.remove(this.CssClasses_.IS_VISIBLE);
-      }.bind(this), 0);
+  MaterialRipple.prototype.upHandler_ = function() {
+    // set mouseup to allow downhanlder to remove visible state
```
Comment (sompylasar) [inline src/ripple/ripple.js]: Please capitalize properly, and terminate with a period: `Set mouseUp to allow downHanlder to remove visible state.`

File: src/ripple/ripple.js (hunk 4 above)
Comment (akotlar) [inline src/ripple/ripple.js]: Those weren't meant to be part of the fix. Would you like me to submit a different pull request, to 1.1?

**Comment by sompylasar** [comment]
@akotlar Great, thanks!
BTW, not my codebase :) just reviewing.

File: src/ripple/ripple.js (hunk 3 above)
Comment (akotlar) [inline src/ripple/ripple.js]: Ok, will update.

File: src/ripple/ripple.js (hunk 1 above)
Comment (akotlar) [inline src/ripple/ripple.js]: Ok, will change to request alone.

File: src/ripple/ripple.js (hunk 5 above)
Comment (akotlar) [inline src/ripple/ripple.js]: Got it thanks!

**Comment by akotlar** [comment]
Thanks for the great feedback! It looks like I need to update a few items, and retry.

File: src/ripple/ripple.js (hunk 3 above)
Comment (akotlar) [inline src/ripple/ripple.js]: John, wanted to say that you have a great attention to detail. I like your idea, since requestAnimationFrame callbacks are executed ~60fps this is certainly important.

File: src/ripple/ripple.js (hunk 2 above)
Comment (akotlar) [inline src/ripple/ripple.js]: Absolutely, to confirm, I SHOULD have merged into mdl-1.1, they use window.setTimeout. Could you tell me, in your/colleagues experience, is setTimeout / requestAnimationFrame better accepted and therefore more readable that window.*? I would think we want to give the full context (i.e to which object this method belongs)? Thank you.

Comment (sompylasar) [inline src/ripple/ripple.js]: In my experience, there are two sides of this coin, neither of which relate to this comment (just to be consistent to the surrounding coding style). One side is shimming the `window` object via webpack imports loader (you can't shim references to free variables, so `window.` is preferred). The other is isomorphic context, where on the server-side there is no window object, but there are functions in the global object (which in the browser equals to the window object).

File: src/ripple/ripple.js (hunk 3 above)
Comment (sompylasar) [inline src/ripple/ripple.js]: Thanks)

**Comment by Garbee** [comment]
Also, this should be off the `mdl-1.1` branch and be pushed into that. `master` is currently v2 experimentation as explained in the README.

**Comment by akotlar** [comment]
Yes, thank you. I've done this. I wasn't sure whether README.md or CONTRIBUTING.md instructions were more recent (contributing specifies master). Thank you both, very much for your help with my first small contribution.
